File: Client/Captions/CaptionsLib/source/srt/SRTUtils.cs
﻿//Filenam: SRTUtils.cs
//Version: 20131114

using ClipFlair.CaptionsLib.Utils;

using System;
using Microsoft.SilverlightMediaFramework.Core.Accessibility.Captions;

namespace ClipFlair.CaptionsLib.SRT
{

  public sealed class SRTUtils
  {

    public const string SRTtimeFormat = "HH:mm:ss,fff";
      //Must use 2 instead of 3 as done at all the ClipFlair.CaptionsLib controls
    public const int SignificantDigits = 2;

    public const string SRT_TIME_SEPARATOR = " --> ";

    public static DateTime BaseTime = DateTimeUtils.DATETIMEZERO;
    public static string SecondsToSRTtime(double seconds)
    {
      return DateTimeUtils.SecondsToDateTimeStr(seconds, BaseTime, SRTtimeFormat, SignificantDigits);
    }

    public static double SRTtimeToSeconds(string srtTime)
    {
      return DateTimeUtils.TimeStrToSeconds(srtTime, BaseTime, SRTtimeFormat, SignificantDigits);
    }

    public static void SRTStringToCaption(string srtString, CaptionElement caption)
    {
      try
      {
        if (srtString != null)
        {
          string[] TimesAndCaptions = StringUtils.Split(srtString, StringUtils.vbCrLf);

          string[] TimesOnly = StringUtils.Split(TimesAndCaptions[1], SRT_TIME_SEPARATOR);
          caption.Begin = TimeSpan.FromSeconds(SRTtimeToSeconds(TimesOnly[0]));
          caption.End = TimeSpan.FromSeconds(SRTtimeToSeconds(TimesOnly[1]));

          caption.Content = "";
          for (int i = 2; i <= TimesAndCaptions.Length - 1; i++)
          {
            if ((caption.Content != ""))
              caption.Content += StringUtils.vbCrLf;
            caption.Content += TimesAndCaptions[i];
          }
        }
      }
      catch
      {
        throw new Exception("Invalid SRT");
        //TODO: localize
      }
    }

  }

}

File: Client/Captions/CaptionsLib/VB/Source/srt/SRTUtils.vb
'Filenam: SRTUtils.vb
'Version: 20121119

Imports ClipFlair.CaptionsLib.Utils.DateTimeUtils

Imports Microsoft.SilverlightMediaFramework.Core.Accessibility.Captions

Namespace ClipFlair.CaptionsLib.SRT

  Public NotInheritable Class SRTUtils

    Public Const SRTtimeFormat As String = "HH:mm:ss,fff"
    Public Const SignificantDigits As Integer = 2 'Must use 2 instead of 3 as done at all the ClipFlair.CaptionsLib controls
    Public Const SRT_TIME_SEPARATOR As String = " --> "

    Public Shared BaseTime As DateTime = DATETIMEZERO

    Public Shared Function SecondsToSRTtime(ByVal seconds As Double) As String
      Return SecondsToDateTimeStr(seconds, BaseTime, SRTtimeFormat, SignificantDigits)
    End Function

    Public Shared Function SRTtimeToSeconds(ByVal srtTime As String) As Double
      Return TimeStrToSeconds(srtTime, BaseTime, SRTtimeFormat, SignificantDigits)
    End Function

    Public Shared Sub SRTStringToCaption(ByVal srtString As String, ByVal caption As CaptionElement)
      Try
        If srtString IsNot Nothing Then
          Dim TimesAndCaptions() As String = Split(srtString, vbCrLf)

          Dim TimesOnly() As String = Split(TimesAndCaptions(1), SRT_TIME_SEPARATOR)
          caption.Begin = TimeSpan.FromSeconds(SRTtimeToSeconds(TimesOnly(0)))
          caption.End = TimeSpan.FromSeconds(SRTtimeToSeconds(TimesOnly(1)))

          caption.Content = ""
          For i As Integer = 2 To TimesAndCaptions.Length - 1
            If (CStr(caption.Content) <> "") Then caption.Content += vbCrLf
            caption.Content += TimesAndCaptions(i)
          Next
        End If
      Catch
        Throw New Exception("Invalid SRT") 'TODO: localize
      End Try
    End Sub

  End Class

End Namespace
